src/Database/Esqueleto/PostgreSQL/JSON/Instances.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Database.Esqueleto.PostgreSQL.JSON.Instances where
+
+import Data.Aeson (FromJSON(..), ToJSON(..), encode, eitherDecodeStrict)
+import Data.Bifunctor (first)
+import qualified Data.ByteString.Lazy as BSL (toStrict)
+import Data.String (IsString(..))
+import Data.Text (Text)
+import qualified Data.Text as T (concat, pack)
+import qualified Data.Text.Encoding as TE (decodeUtf8, encodeUtf8)
+import Database.Esqueleto (Value, just, val)
+import Database.Esqueleto.Internal.PersistentImport
+import Database.Esqueleto.Internal.Sql (SqlExpr)
+import GHC.Generics (Generic)
+
+
+-- | Newtype wrapper around any type with a JSON representation.
+--
+-- @since 3.1.0
+newtype JSONB a = JSONB { unJSONB :: a }
+  deriving
+    ( Generic
+    , FromJSON
+    , ToJSON
+    , Eq
+    , Foldable
+    , Functor
+    , Ord
+    , Read
+    , Show
+    , Traversable
+    )
+
+-- | 'SqlExpr' of a NULL-able 'JSONB' value.
+type JSONExpr a = SqlExpr (Value (Maybe (JSONB a)))
```

Comment:
One small point - I understand why the `Maybe` is necessary here, but I find it a bit strange that `JSONExpr` is the name of the synonym for nullable JSONB values. 

When I read `(->.) :: JSONExpr a -> JSONAccessor -> JSONExpr b`, I don't see any immediate indication that anything there is nullable, and I have to look at the type synonym to see it.

I'd suggest changing the name to indicate the `Maybe` in there - perhaps:

```haskell
-- | 'SqlExpr' of a NULL-able 'JSONB' value.
-- Note: The null here is a SQL null, not a JSON null.
type JSONMaybeExpr a = SqlExpr (Value (Maybe (JSONB a))) -- name isn't great
```

On a related note, I don't see any way to operate on _non-nullable_ JSONB values. Is that by design? Presumably it would be quite hard to provide an API which dealt with both without a lot of duplication, and I suppose I can always use `just` on any exprs from non-nullable columns.

Please don't consider this to be blocking at all.

Reply:
On the subject of changing the name, maybe it would also be good to add the `B` to indicate this in `JSONB`, in case there is any chance of `JSON` types being added later.

Reply:
I'd say, you'll notice when you use it, and you can always look at the type synonym.
If someone has a good suggestion on how to make this clearer, I can easily integrate it.

Reply:
> On the subject of changing the name, maybe it would also be good to add the `B` to indicate this in `JSONB`, in case there is any chance of `JSON` types being added later.

This is probably a good idea, so I'll adjust that.